tsdb/shard.go

```diff
@@ -43,6 +43,25 @@ var (
 	ErrFieldUnmappedID = errors.New("field ID not mapped")
 )
 
+// A ShardError implements the error interface, and contains extra
+// context about the shard that generated the error.
+type ShardError struct {
+	id  uint64
+	Err error
+}
+
+// NewShardError returns a new ShardError.
+func NewShardError(id uint64, err error) ShardError {
+	if err == nil {
+		err = errors.New("unknown error")
```

Comment:
Yeah makes sense, though I'm not sure it makes sense to get `nil` returned from `NewShardError`. I wonder if it might make sense to just return a `ShardError` and adjust the `Sprintf` in `Error()` so that it returns something like `[shard 3] <nil>`.